pkg/serviceaccount/legacy.go

```diff
@@ -96,6 +107,27 @@ func (v *legacyValidator) Validate(ctx context.Context, tokenData string, public
 		return nil, errors.New("sub claim is invalid")
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LegacyServiceAccountTokenTracking) {
+		warning.AddWarning(ctx, "", "Legacy token was used. Consider switching to [TokenRequest](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/) API.")
+		secret, err := v.getter.GetSecret(namespace, secretName)
+		if err != nil {
+			klog.V(4).Infof("Could not retrieve token %s/%s for service account %s/%s: %v", namespace, secretName, namespace, serviceAccountName, err)
+			return nil, errors.New("Token has been invalidated")
+		}
+		annotations := secret.GetAnnotations()
+		today := time.Now().Format("2006-01-02")
+		if t, ok := annotations[LastUsedAnnotationKey]; !ok || t != today {
+			annotations[LastUsedAnnotationKey] = today
+			secret.SetAnnotations(annotations)
+			labels := secret.GetLabels()
+			labels[InUsedLabelKey] = today
+			secret.SetLabels(labels)
+			if _, err := v.secretsGetter.Secrets(namespace).Update(ctx, secret, metav1.UpdateOptions{}); err != nil {
+				klog.Errorf("Failed to annotate legacy service account token secret with last-used, err: %v", err)
+			}
```

Comment:
Submit a merge patch so that concurrent updates won't conflict. Something like:

```go
			patchContent, _ := json.Marshal(corev1apply.
				Secret(secret.Name, secret.Namespace).
				WithLabels(map[string]string{LastUsedAnnotationKey: today}).
				WithAnnotations(map[string]string{InUsedLabelKey: today}))
			if _, err := v.secretsGetter.Secrets(namespace).Patch(ctx, secret.Name, types.MergePatchType, patchContent, metav1.PatchOptions{}); err != nil {
				klog.Errorf("Failed to annotate legacy service account token secret with last-used, err: %v", err)
			}
```

Reply:
I'd also suggest silencing the error if it is a NotFound error... if the secret is gone then we don't need to report a failure to annotate it

For other types of errors, I'm wondering if we need a metric or some indicator that we're having trouble recording the last used dates... we wouldn't want to start cleaning up tokens we thought were unused if we knew there was trouble recording last-used dates

Reply:
recording a metric will have the same problem as we don't know if it succeeded, which would need another metric....  would retry + one year gap between tracking and clean up sufficient?